File: 4 kyu/Human readable duration format/solution.jl
function formatduration(seconds)
  if seconds < 1
    return "now"
  end
  formatted = []
  times = [
    "year" => 31536000,
    "day" => 86400,
    "hour" => 3600,
    "minute" => 60,
    "second" => 1
  ]
  for t in times
    s = Int(floor(seconds / t[2]))
    if s >= 1
      push!(formatted, string(s) * " " * t[1] * (s > 1 ? "s" : ""))
      seconds %= t[2]
    end
  end
  return join(formatted, ", ", " and ")
end

__________________________________________________
function compose_sentence(arr)
    if length(arr) <= 1
        return join(arr)
    else
        start = join(arr[1:end-1], ", ")
        return "$(start) and $(arr[end])"
    end
end

function formatduration(s)
    if s == 0
        return "now"
    end
    NAMES = ["year", "day", "hour", "minute", "second"]
    y, s = fldmod(s, 31536000)
    d, s = fldmod(s, 86400)
    h, s = fldmod(s, 3600)
    m, s = fldmod(s, 60)
    res = ["$(num) $(st)$('s' ^ (num > 1))" for (num, st) in zip([y, d, h, m, s], NAMES) if num >= 1]
    return compose_sentence(res)
end

__________________________________________________
function formatduration(s)
  if s==0
    return "now"
  end
  v = [s÷31536000, s÷86400%365, s÷3600%24, s÷60%60, s%60]
  strs = ["$(v[1]) year", "$(v[2]) day", "$(v[3]) hour", "$(v[4]) minute", "$(v[5]) second"]
  join(collect(string(strs[i],v[i]!=1 ? "s" : "") for i in 1:5 if v[i]!=0), ", ", " and ")
end

__________________________________________________
function formatduration(t)
    iszero(t) && return "now"

    m, s = divrem(t, 60)
    h, m = divrem(m, 60)
    d, h = divrem(h, 24)
    y, d = divrem(d, 365)

    result = iszero(s) ? "" : "-$(s) second$(s > 1 ? "s" : "")"
    result *= iszero(m) ? "" : "-$(m) minute$(m > 1 ? "s" : "")"
    result *= iszero(h) ? "" : "-$(h) hour$(h > 1 ? "s" : "")"
    result *= iszero(d) ? "" : "-$(d) day$(d > 1 ? "s" : "")"
    result *= iszero(y) ? "" : "-$(y) year$(y > 1 ? "s" : "")"
    
    return(join(reverse(split(result, '-')[2:end]), ", ", " and "))
end

__________________________________________________
function formatduration(time)
    time == 0 && return "now"

    minutes, seconds = divrem(time, 60)
    hours, minutes = divrem(minutes, 60)
    days, hours = divrem(hours, 24)
    years, days = divrem(days, 365)

    t = (y=years, d=days, h=hours, m=minutes, s=seconds)

    result = iszero(t.s) ? "" : "-$(t.s) second$(t.s > 1 ? "s" : "")"
    result *= iszero(t.m) ? "" : "-$(t.m) minute$(t.m > 1 ? "s" : "")"
    result *= iszero(t.h) ? "" : "-$(t.h) hour$(t.h > 1 ? "s" : "")"
    result *= iszero(t.d) ? "" : "-$(t.d) day$(t.d > 1 ? "s" : "")"
    result *= iszero(t.y) ? "" : "-$(t.y) year$(t.y > 1 ? "s" : "")"
    
    return(join(reverse(split(result, '-')[2:end]), ", ", " and "))
end

__________________________________________________
function formatduration(seconds)
    if iszero(seconds)
        return "now"
    end
    format = ["year", "day", "hour", "minute", "second"]
    y, s = divrem(seconds, 31536000)
    d, s = divrem(s, 86400)
    h, s = divrem(s, 3600)
    m, s = divrem(s, 60)
    date = (y, d, h, m, s)
    nvalid = count(!iszero, date)
    str = ""
    for (i, x) in enumerate(date)
        if iszero(x)
            continue
        end
        str *= string(x, " ", format[i], !isone(x) ? "s" : "")
        nvalid -= 1
        if nvalid > 1
            str *= ", "
        elseif nvalid == 1
            str *= " and "
        end
    end
    str
end

File: 4 kyu/Human readable duration format/solution.py
times = [("year", 365 * 24 * 60 * 60), 
         ("day", 24 * 60 * 60),
         ("hour", 60 * 60),
         ("minute", 60),
         ("second", 1)]

def format_duration(seconds):

    if not seconds:
        return "now"

    chunks = []
    for name, secs in times:
        qty = seconds // secs
        if qty:
            if qty > 1:
                name += "s"
            chunks.append(str(qty) + " " + name)

        seconds = seconds % secs

    return ', '.join(chunks[:-1]) + ' and ' + chunks[-1] if len(chunks) > 1 else chunks[0]
  
___________________________________________________
def format_duration(seconds):
    if seconds == 0: return "now"
    units = ( (31536000, "year"  ), 
              (   86400, "day"   ),
              (    3600, "hour"  ),
              (      60, "minute"),
              (       1, "second") )
    ts, t = [], seconds
    for unit in units:
        u, t = divmod(t, unit[0])
        ts += ["{} {}{}".format(u, unit[1], "s" if u>1 else "")] if u != 0 else []
    return ", ".join([str(d)for d in ts[:-1]]) + (" and " if len(ts)>1 else "") + ts[-1]
  
___________________________________________________
def format_duration(total_sec):
    years = int(total_sec / 31536000)
    days = int(total_sec / 86400 % 365)
    hours = int(total_sec / 3600 % 24)
    minutes = int(total_sec / 60 % 60)
    seconds = int(total_sec % 60)
    date = ''
    if years >= 1:
        if years == 1:
            date += str(years) + ' year'
        else:
            date += str(years) + ' years'
    if days >= 1:
        if days == 1:
            if date == '':
                date += str(days) + ' day'
            elif hours == minutes == seconds ==0:
                date +=' and ' + str(days) + ' day'
            else:
                date +=', ' + str(days) + ' day'
        else:
            if date == '':
                date += str(days) + ' days'
            elif hours == minutes == seconds ==0:
                date +=' and ' + str(days) + ' days'
            else:
                date +=', ' + str(days) + ' days'
    if hours >= 1:
        if hours == 1:
            if date == '':
                date += str(hours) + ' hour'
            elif minutes == seconds ==0:
                date +=' and ' + str(hours) + ' hour'
            else:
                date +=', ' + str(hours) + ' hour'
        else:
            if date == '':
                date += str(hours) + ' hours'
            elif minutes == seconds ==0:
                date +=' and ' + str(hours) + ' hours'
            else:
                date +=', ' + str(hours) + ' hours'
    if minutes >= 1:
        if minutes == 1:
            if date == '':
                date += str(minutes) + ' minute'
            elif seconds ==0:
                date +=' and ' + str(minutes) + ' minute'
            else:
                date +=', ' + str(minutes) + ' minute'
        else:
            if date == '':
                date += str(minutes) + ' minutes'
            elif seconds ==0:
                date +=' and ' + str(minutes) + ' minutes'
            else:
                date += ', ' + str(minutes) + ' minutes'
    if seconds >= 1:
        if seconds == 1:
            if date == '':
                date += str(seconds) + ' second'
            else:
                date += ' and ' + str(seconds) + ' second'
        else:
            if date == '':
                date += str(seconds) + ' seconds'
            else:
                date += ' and ' + str(seconds) + ' seconds'
    if date == '':
        return ('now')
    else:
        return(date)
      
___________________________________________________
def format_duration(seconds):
    if seconds == 0:
        return 'now'
    years,rest = divmod(seconds, 3600*24*365)
    days, rest = divmod(rest, 3600*24)
    hours, rest = divmod(rest, 3600)
    minutes, seconds = divmod(rest, 60)
    final_arr = []
    if seconds > 0: 
        final_arr.append(f"{seconds} second{'s' if seconds > 1 else ''}")
    if minutes > 0:
        minutes_string = final_arr.append(f"{minutes} minute{'s' if minutes > 1 else ''}")
    if hours > 0:    
        final_arr.append(f"{hours} hour{'s' if hours > 1 else ''}")
    if days > 0:
        final_arr.append(f"{days} day{'s' if days > 1 else ''}")
    if years > 0:
        final_arr.append(f"{years} year{'s' if years > 1 else ''}")
    if len(final_arr) == 1:
        return final_arr[0]
    elif len(final_arr) == 2:
        final_arr.reverse()
        return " and ".join(final_arr)
    elif len(final_arr) == 3:
        return f"{final_arr[2]}, {final_arr[1]} and {final_arr[0]}"
    elif len(final_arr) == 4:
        return f"{final_arr[3]}, {final_arr[2]}, {final_arr[1]} and {final_arr[0]}"
    else:
        return f"{final_arr[4]}, {final_arr[3]}, {final_arr[2]}, {final_arr[1]} and {final_arr[0]}"
        
___________________________________________________
def format_duration(seconds):
    if seconds == 0:
        return "now"
    
    divisors = [31_536_000, 86_400, 3600, 60, 1]
    unit_names = ['year', 'day', 'hour', 'minute', 'second']
    unit_quants = []
    
    for i in range(5):
        quant = seconds // divisors[i]
        seconds = seconds % divisors[i]
        if quant == 1:
            unit_quants.append(f'1 {unit_names[i]}')
        elif quant:
            unit_quants.append(f'{quant} {unit_names[i]}s')
        
    return (", ".join(unit_quants[:-1]) +
            (" and " if len(unit_quants) > 1 else "") + unit_quants[-1])
    
___________________________________________________
def format_duration(seconds):
    words = ["year", "day", "hour", "minute", "second"]
    if seconds == 0:
        return "now"
    else:
        m, s = divmod(seconds, 60)
        h, m = divmod(m, 60)
        d, h = divmod(h, 24)
        y, d = divmod(d, 365)
        human = [y, d, h, m, s]
        duration = []
        for x, i in enumerate(human):
            if i == 1:
                duration.append(f"{i} {words[x]}")
            elif i > 1:
                duration.append(f"{i} {words[x]}s")

        if len(duration) == 1:
            return duration[0]
        elif len(duration) == 2:
            return f"{duration[0]} and {duration[1]}"
        else:
            return ", ".join(duration[:-1]) + " and " + duration[-1]
